api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go

```diff
@@ -0,0 +1,212 @@
+	ctx, span := observability.StartSpan("Piccolo Target Provider", ctx, &map[string]string{
+		"method": "Get",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Infof("  P (Piccolo Target): getting artifacts: %s - %s, traceId: %s", deployment.Instance.Spec.Scope, deployment.Instance.ObjectMeta.Name, span.SpanContext().TraceID().String())
+
+	ret := make([]model.ComponentSpec, 0)
+	for _, component := range references {
+		name := component.Component.Name
+		req, err := http.NewRequest("GET", "http://0.0.0.0:9090/scenario/"+name, nil)
+		if err != nil {
+			continue
+		}
+		client := &http.Client{}
+		resp, err := client.Do(req)
+
+		if err != nil {
+			continue
+		}
+
+		defer resp.Body.Close()
+
+		switch resp.StatusCode {
+		case http.StatusOK:
+			// respBody, err := io.ReadAll(resp.Body)
+			component := model.ComponentSpec{
+				Name:       name,
+				Properties: make(map[string]interface{}),
+			}
+			component.Properties["workload.name"] = string(name)
+			ret = append(ret, component)
+		case http.StatusNotFound:
+			continue
+		}
+	}
+
+	return ret, nil
+}
+
+func (i *PiccoloTargetProvider) Apply(ctx context.Context, deployment model.DeploymentSpec, step model.DeploymentStep, isDryRun bool) (map[string]model.ComponentResultSpec, error) {
+	ctx, span := observability.StartSpan("Piccolo Target Provider", ctx, &map[string]string{
+		"method": "Apply",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Infof("  P (Piccolo Target): applying artifacts: %s - %s, traceId: %s", deployment.Instance.Spec.Scope, deployment.Instance.ObjectMeta.Name, span.SpanContext().TraceID().String())
+
+	injections := &model.ValueInjections{
+		InstanceId: deployment.Instance.ObjectMeta.Name,
+		SolutionId: deployment.Instance.Spec.Solution,
+		TargetId:   deployment.ActiveTarget,
+	}
+
+	components := step.GetComponents()
+	err = i.GetValidationRule(ctx).Validate(components)
+	if err != nil {
+		sLog.Errorf("  P (Piccolo Target): failed to validate components: %+v, traceId: %s", err, span.SpanContext().TraceID().String())
+		return nil, err
+	}
+	if isDryRun {
+		err = nil
+		return nil, nil
+	}
+
+	ret := step.PrepareResultMap()
+
+	for _, component := range step.Components {
+		if component.Action == model.ComponentUpdate {
+			name := model.ReadPropertyCompat(component.Component.Properties, "workload.name", injections)
+			if name == "" {
+				err = errors.New("component doesn't have workload.name property")
+				ret[component.Component.Name] = model.ComponentResultSpec{
+					Status:  v1alpha2.UpdateFailed,
+					Message: err.Error(),
+				}
+				sLog.Errorf("  P (Piccolo Target): %+v, traceId: %s", err, span.SpanContext().TraceID().String())
+				return ret, err
+			}
+			resp, err := http.Post("http://0.0.0.0:9090/create-scenario/"+component.Component.Name, "text/plain", nil)
+			if err != nil {
+				sLog.Errorf("  P (Piccolo Target): fail to create resource")
```

Comment:
sLog.ErrorfCtx

Reply:
This commit was written too long ago, so the ErrorfCtx method does not exist.
I will create a new PR later to fix this.